src/repr/src/row.rs

```diff
@@ -375,8 +376,20 @@ unsafe fn read_datum<'a>(data: &'a [u8], offset: &mut usize) -> Datum<'a> {
         Tag::JsonNull => Datum::JsonNull,
         Tag::Dummy => Datum::Dummy,
         Tag::APD => {
-            let n = read_copy::<OrderedDecimal<Apd>>(data, offset);
-            Datum::APD(n)
+            let digits = read_copy::<u8>(data, offset);
+            let exponent = read_copy::<i8>(data, offset);
+            let bits = read_copy::<u8>(data, offset);
+            let lsu = read_untagged_bytes(data, offset);
+            // SAFETY: `lsu`'s data originated as a `&[u16]`, before being
+            // transmuted into a `&[u8]`; this transmute returns it to its
+            // original form.
+            let (prefix, lsu, suffix) = lsu.align_to::<u16>();
+            // LSU should have essentially ceil(digits / 3) elements.
+            assert!(lsu.len() == (usize::try_from(digits).unwrap() + 2) / 3);
```

Comment:
I am afraid I am hitting this assertion:

```
CREATE TABLE t1 (f1 APD) ;
SELECT COUNT(*)::TEXT::APD AS f1 FROM t1;
```

```
materialized v0.7.4-dev (a312dde63) listening on 0.0.0.0:6875...
2021-06-07T06:36:44.848484Z ERROR panic: assertion failed: lsu.len() == (usize::try_from(digits).unwrap() + 2) / 3
thread: timely:work-0
location: src/repr/src/row.rs:388:13
version: 0.7.4-dev (a312dde635788108186d4407d1ab12fb261fb659)
backtrace:
   0: materialized::handle_panic
             at src/materialized/src/bin/materialized/main.rs:710:21
   1: core::ops::function::Fn::call
             at /rustc/2fd73fabe469357a12c2c974c140f67e7cdd76d0/library/core/src/ops/function.rs:70:5
   2: std::panicking::rust_panic_with_hook
             at /rustc/2fd73fabe469357a12c2c974c140f67e7cdd76d0/library/std/src/panicking.rs:595:17
   3: std::panicking::begin_panic_handler::{{closure}}
             at /rustc/2fd73fabe469357a12c2c974c140f67e7cdd76d0/library/std/src/panicking.rs:495:13
   4: std::sys_common::backtrace::__rust_end_short_backtrace
             at /rustc/2fd73fabe469357a12c2c974c140f67e7cdd76d0/library/std/src/sys_common/backtrace.rs:141:18
   5: rust_begin_unwind
             at /rustc/2fd73fabe469357a12c2c974c140f67e7cdd76d0/library/std/src/panicking.rs:493:5
   6: core::panicking::panic_fmt
             at /rustc/2fd73fabe469357a12c2c974c140f67e7cdd76d0/library/core/src/panicking.rs:92:14
   7: core::panicking::panic
             at /rustc/2fd73fabe469357a12c2c974c140f67e7cdd76d0/library/core/src/panicking.rs:50:5
   8: repr::row::read_datum
             at src/repr/src/row.rs:388:13
   9: <repr::row::DatumListIter as core::iter::traits::iterator::Iterator>::next
             at src/repr/src/row.rs:1121:27
      core::iter::traits::iterator::Iterator::fold
             at /rustc/2fd73fabe469357a12c2c974c140f67e7cdd76d0/library/core/src/iter/traits/iterator.rs:2123:29
      core::iter::traits::iterator::Iterator::count
             at /rustc/2fd73fabe469357a12c2c974c140f67e7cdd76d0/library/core/src/iter/traits/iterator.rs:251:9
      repr::row::Row::unpack
             at src/repr/src/row.rs:765:19
  10: dataflow::render::<impl dataflow::render::context::Context<G,expr::relation::MirRelationExpr,repr::row::Row,u64>>::ensure_rendered::{{closure}}
             at src/dataflow/src/render/mod.rs:602:46
      differential_dataflow::collection::Collection<G,D,R>::map::{{closure}}
```

Reply:
This PR now includes MaterializeInc/rust-dec#42, which obviated the need for this assertion, but seems to have resolved the underlying issue that caused this panic